segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/StreamSegmentMapper.java

```diff
@@ -515,9 +560,61 @@ private boolean isValidStreamSegmentId(long id) {
 
     //endregion
 
+    //region Helper Classes
+
     @Data
     private static class SegmentInfo {
         private final long segmentId;
         private final SegmentProperties properties;
     }
+
+    /**
+     * A pending request for a Segment Assignment, which keeps track of all queued callbacks.
+     * Note that this class in itself is not thread safe, so the caller should take precautions to ensure thread safety.
+     */
+    private static class PendingRequest {
+        private final ArrayList<QueuedCallback<?>> callbacks = new ArrayList<>();
+
+        /**
+         * Invokes all queued callbacks, in order, with the given SegmentId as a parameter.
+         */
+        void complete(long segmentId) {
+            for (QueuedCallback<?> callback : this.callbacks) {
+                try {
+                    callback.complete(segmentId);
```

Comment:
This strikes me as a bit dangrous. It is expected that the caller is holding a lock because this class is not thread safe and a callback is being invoked. This could extend the duration which the lock is held rather dramatically. It might be better to make this thread safe so that you can ensure that a lock is not held when the callback is being invoked.